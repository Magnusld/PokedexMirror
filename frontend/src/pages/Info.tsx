import React, { useEffect, useState } from 'react'
import { Button, Card } from 'react-bootstrap';
import '../style/Info.css';
import pokemon from "../images/376.jpg";
import { Link, useParams } from 'react-router-dom';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import { PokemonAdvanced } from '../types';
import PokemonStarRating from "../components/PokemonStarRating";
import PokemonAggregatedRating from "../components/PokemonAggregatedRating";
import { getFullSize } from '../util/imageResolver';

import {ReactComponent as Bug} from "../assets/typeIconsSvg/Pokémon_Bug_Type_Icon.svg";
import {ReactComponent as Dark} from "../assets/typeIconsSvg/Pokémon_Dark_Type_Icon.svg";
import {ReactComponent as Dragon} from "../assets/typeIconsSvg/Pokémon_Dragon_Type_Icon.svg";
import {ReactComponent as Electric} from "../assets/typeIconsSvg/Pokémon_Electric_Type_Icon.svg";
import {ReactComponent as Fairy} from "../assets/typeIconsSvg/Pokémon_Fairy_Type_Icon.svg";
import {ReactComponent as Fighting} from "../assets/typeIconsSvg/Pokémon_Fighting_Type_Icon.svg";
import {ReactComponent as Fire} from "../assets/typeIconsSvg/Pokémon_Fire_Type_Icon.svg";
import {ReactComponent as Flying} from "../assets/typeIconsSvg/Pokémon_Flying_Type_Icon.svg";
import {ReactComponent as Ghost} from "../assets/typeIconsSvg/Pokémon_Ghost_Type_Icon.svg";
import {ReactComponent as Grass} from "../assets/typeIconsSvg/Pokémon_Grass_Type_Icon.svg";
import {ReactComponent as Ground} from "../assets/typeIconsSvg/Pokémon_Ground_Type_Icon.svg";
import {ReactComponent as Ice} from "../assets/typeIconsSvg/Pokémon_Ice_Type_Icon.svg";
import {ReactComponent as Normal} from "../assets/typeIconsSvg/Pokémon_Normal_Type_Icon.svg";
import {ReactComponent as Poison} from "../assets/typeIconsSvg/Pokémon_Poison_Type_Icon.svg";
import {ReactComponent as Psychic} from "../assets/typeIconsSvg/Pokémon_Psychic_Type_Icon.svg";
import {ReactComponent as Rock} from "../assets/typeIconsSvg/Pokémon_Rock_Type_Icon.svg";
import {ReactComponent as Steel} from "../assets/typeIconsSvg/Pokémon_Steel_Type_Icon.svg";
import {ReactComponent as Water} from "../assets/typeIconsSvg/Pokémon_Water_Type_Icon.svg";

function Info() {

    type PokemonParams = {
        id: string;
    };      

    interface PokemonAdvancedData {
        pokemon: PokemonAdvanced
    }

    const { id } = useParams<PokemonParams>()

    const GET_POKEMON_INFO = gql`
    query($where: PokemonWhereUniqueInput!) {
        pokemon(where: $where) {
          id
          pokedexNr
          name
          generation
          species
          type1
          heightMeter
          type2
          weightKg
          hp
          attack
          defense
          sp_attack
          sp_defense
          speed
          ability1
          ability2
          ability3
          aggregated_rating
        }
      }
    `;

    const variables = {
        variables: {
            "where": {
                "id": parseInt(id)
            }
        }  
    }

    const { loading, error, data } = useQuery<PokemonAdvancedData, any>(GET_POKEMON_INFO, variables);

    const [hp, setHp] = useState<number>(0);
    const [attack, setAttack] = useState<number>(0);
    const [defense, setDefense] = useState<number>(0);
    const [spAttack, setSpAttack] = useState<number>(0);
    const [spDefense, setSpDefense] = useState<number>(0);
    const [speed, setSpeed] = useState<number>(0);

    useEffect(() => {
        console.log("useEffect")
        if(data?.pokemon) setHp(data?.pokemon.hp)
        if(data?.pokemon) setAttack(data?.pokemon.attack)
        if(data?.pokemon) setDefense(data?.pokemon.defense)
        if(data?.pokemon) setSpAttack(data?.pokemon.sp_attack)
        if(data?.pokemon) setSpDefense(data?.pokemon.sp_defense)
        if(data?.pokemon) setSpeed(data?.pokemon.speed)
    }, [data])

    /**
     * takes potential type 1 and type 2 and makes it into a list
     * @param types rest param with one or two types
     * @returns 
     */
    function typesToList(...types : string[]) {
        const typeList = types.filter((type) => type.length > 0);
        return typeList;
    }

    let showTypeEmblem = typesToList(data ? data.pokemon.type1 : "", data ? data.pokemon.type2 : "").map((Type) => (
        <span>
          {Type === "Bug" ? <Bug className={"typeIcon"}/> : null}
          {Type === "Dark" ? <Dark className={"typeIcon"}/> : null}
          {Type === "Dragon" ? <Dragon className={"typeIcon"}/> : null}
          {Type === "Electric" ? <Electric className={"typeIcon"}/> : null}
          {Type === "Fairy" ? <Fairy className={"typeIcon"}/> : null}
          {Type === "Fighting" ? <Fighting className={"typeIcon"}/> : null}
          {Type === "Fire" ? <Fire className={"typeIcon"}/> : null}
          {Type === "Flying" ? <Flying className={"typeIcon"}/> : null}
          {Type === "Ghost" ? <Ghost className={"typeIcon"}/> : null}
          {Type === "Grass" ? <Grass className={"typeIcon"}/> : null}
          {Type === "Ground" ? <Ground className={"typeIcon"}/> : null}
          {Type === "Ice" ? <Ice className={"typeIcon"}/> : null}
          {Type === "Normal" ? <Normal className={"typeIcon"}/> : null}
          {Type === "Poison" ? <Poison className={"typeIcon"}/> : null}
          {Type === "Psychic" ? <Psychic className={"typeIcon"}/> : null}
          {Type === "Rock" ? <Rock className={"typeIcon"}/> : null}
          {Type === "Steel" ? <Steel className={"typeIcon"}/> : null}
          {Type === "Water" ? <Water className={"typeIcon"}/> : null}
        </span>
    ))

    function generateTypes(types : string[]) {
        if(types.length === 1) {
            let color1 = setTypeColor(types[0])
            return <div className="types"><span className="type-icon" style={{backgroundColor: color1}}>{types[0]}</span></div>
        }
        else if(types.length === 2) {
            let color1 = setTypeColor(types[0])
            let color2 = setTypeColor(types[1])
            return <div className="types"><span className="type-icon" style={{backgroundColor: color1}}>{types[0]}</span><span className="type-icon" style={{backgroundColor: color2}}>{types[1]}</span></div>
        }
        else {
            return <div className="types"><span className="type-icon" style={{backgroundColor: "white"}}>None</span></div>
        }
    }

    function generateAbilities(ability1: string, ability2: string, ability3: string) {
        return <div className="ability-list">
            {ability1.length > 0 ? <span>1. {ability1}</span> : null}
            {ability2.length > 0 ? <span>2. {ability2}</span> : null}
            {ability3.length > 0 ? <span style={{color: "orange"}}>(Hidden ability) {ability3}</span> : null}
        </div>
    }

    const colours : any = {
        normal: '#A8A77A',
        fire: '#EE8130',
        water: '#6390F0',
        electric: '#F7D02C',
        grass: '#7AC74C',
        ice: '#96D9D6',
        fighting: '#C22E28',
        poison: '#A33EA1',
        ground: '#E2BF65',
        flying: '#A98FF3',
        psychic: '#F95587',
        bug: '#A6B91A',
        rock: '#B6A136',
        ghost: '#735797',
        dragon: '#6F35FC',
        dark: '#705746',
        steel: '#B7B7CE',
        fairy: '#D685AD',
    };

    function setTypeColor(type : string) {
        let string = type.toLowerCase()
        if(string in colours) {
            return colours[string]
        }
        else {
            return 'grey'
        }
        
    }

    console.log(data)

    return (
        <div className={"container"}>
            <header className="info-header-container">
                <Link to="/" className="back-button-link">
                    <div className="back-button-container">
                        <span className="material-icons md-36">arrow_back</span>
                        <span className="back-button-text">Tilbake</span>
                    </div>
                </Link>
            </header>
            <div className="info-container">
                <div className="top-container">
                    <div className="image">
                        <img src={getFullSize(data ? data.pokemon.pokedexNr : 0)} alt={data?.pokemon.name + " bilde"}></img>
                    </div>
                    <div className="basic-info">
                        <div className="name-number">
                            <span className="pokename">{data?.pokemon.name}</span>
                            <span className="pokenum">#{data?.pokemon.pokedexNr}</span>
                        </div>
                        <div className="more-info">
                            <span>{data?.pokemon.species}</span>
                            <span>Generasjon: {data?.pokemon.generation}</span>
                        </div>
                        <div className="types">
                            {showTypeEmblem}
                        </div>
                        <div className={"all-ratings-container"}>
                            <PokemonStarRating pokemonId={parseInt(id)}/>
                            <PokemonAggregatedRating rating={data ? data.pokemon.aggregated_rating : 0}/>
                        </div>
                    </div>
                </div>
                <div className="middle-container">
                    <span>Weight: {data?.pokemon.weightKg}kg</span>
                    <span>Height: {data?.pokemon.heightMeter}m</span>
                </div>
                <div className="bottom-container">
                    <div className="stats">
                        <h2>Stats</h2>
                        <table className="stats-table">
                            <tr>
                                <td>HP</td>
                                <td>{hp}</td>
                            </tr>
                            <tr>
                                <td>Attack</td>
                                <td>{attack}</td>
                            </tr>
                            <tr>
                                <td>Defense</td>
                                <td>{defense}</td>
                            </tr>
                            <tr>
                                <td>Sp. Attack</td>
                                <td>{spAttack}</td>
                            </tr>
                            <tr>
                                <td>Sp. Defense</td>
                                <td>{spDefense}</td>
                            </tr>
                            <tr>
                                <td>Speed</td>
                                <td>{speed}</td>
                            </tr>
                            <tr>
                                <td>Totalt</td>
                                <td>{hp + attack + defense + spAttack + spDefense + speed}</td>
                            </tr>
                        </table>
                    </div>
                    <div className="abilities">
                        <h2>Abilities</h2>
                        {generateAbilities(data ? data.pokemon.ability1 : "", data ? data.pokemon.ability2 : "", data ? data.pokemon.ability3 : "")}
                    </div>
                </div>
            </div>
        </div>

    )
}

export default Info
