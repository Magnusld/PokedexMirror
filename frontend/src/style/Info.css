.info-container {
    --info-container-bg: rgba(3, 101, 140, 1);
    --info-container-boxes: white;
    --pokenum-bg: rgba(3, 76, 140, 1);
}

/* -- FARGE TEMA -- */

/* Color Theme Swatches in Hex */
.Color-Theme-1-hex { color: #FFFFFF; }
.Color-Theme-2-hex { color: #CCFFEB; }
.Color-Theme-3-hex { color: #04BEC4; }
.Color-Theme-4-hex { color: #03658C; }
.Color-Theme-5-hex { color: #034C8C; }

/* Color Theme Swatches in RGBA */
.Color-Theme-1-rgba { color: rgba(255, 255, 255, 1); }
.Color-Theme-2-rgba { color: rgba(204, 255, 235, 1); }
.Color-Theme-3-rgba { color: rgba(4, 190, 196, 1); }
.Color-Theme-4-rgba { color: rgba(3, 101, 140, 1); }
.Color-Theme-5-rgba { color: rgba(3, 76, 140, 1); }

.info-container {
    width: 70vw;
    padding: 10px;
    margin: 20px auto;
    background-color: var(--info-container-bg);
}

.top-container {
    display: grid;
    grid-template-columns: 1fr 2fr;
    grid-gap: 10px;
    height: 300px;
}

.image {
    background-color: var(--info-container-boxes);
    padding: 20px;
}

.basic-info {
    background-color: var(--info-container-boxes);
    padding: 10px;
}

.name-number {
    display: flex;
    align-items: center;
    justify-content: space-between;
    width: 100%;
}

.pokename {
    font-size: 2em;

}

.pokenum {
    text-align: center;
    color: white;
    font-size: 1.2em;
    height: 3em;
    width: 3em;
    background-color: var(--pokenum-bg);
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;

}

.more-info {
    display: flex;
    align-items: center;
    justify-content:flex-start;
    width: 100%;
}

.more-info span:last-of-type {
    margin-left: 30px;
}

.types {
    margin-top: 20px;
    display: flex;
    align-items: center;
    justify-content: flex-start;
    width: 100%;
}

.type-icon {
    padding: 5px 20px;
    font-size: 1.4em;
    font-weight: 600;
    border-radius: 5px;
    border: 1px solid black
}

.type-icon:last-of-type {
    margin-left: 10px;
}

.middle-container {
    width: 100%;
    margin: 10px auto;
    background-color: var(--info-container-boxes);
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 1.4em;
    font-weight: 400;
}

.middle-container span {
    padding: 35px;
}

.bottom-container {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 10px;
    height: 400px;
}

.stats {
    background-color: var(--info-container-boxes);
    padding: 20px 30px;
    font-size: 1.2em;
    font-weight: 500;
}

.stats-table {
    background-color: rgb(248, 248, 248);
    width: 100%;
}

.stats-table tr, .stats-table td {
    border-bottom: solid lightgrey 1px;
    padding: 8px 20px;
}

.stats-table tr:last-of-type, .stats-table td {
    border-bottom: none;
}

.stats-table tr:last-of-type {
    background-color: var(--pokenum-bg);
    color: white;
}

.stats-table td:nth-of-type(2) {
    text-align: right;
}

.stats-column {
    display: flex;
    flex-direction: column;
}

.abilities {
    background-color: var(--info-container-boxes);
    padding: 20px 30px;
    font-size: 1.2em;
    font-weight: 500;
}

.ability-list {
    display: flex;
    flex-direction: column;
}

.ability-list span {
    padding-bottom: 10px;
    font-size: 1em;
    font-weight: 400;
}

.all-ratings-container {
    margin-top: 40px;
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    flex-wrap: wrap;
    gap: 20px;
}

.personal-rating-container {

}

.aggregated-rating-container {

}

@media only screen and (max-width: 900px) {
    .top-container {
        grid-auto-flow: row;
        grid-template-columns: 1fr;
        height: fit-content;
    }

    .bottom-container {
        grid-auto-flow: row;
        grid-template-columns: 1fr;
        height: fit-content;
    }
}