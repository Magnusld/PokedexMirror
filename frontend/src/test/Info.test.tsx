import {render, screen, waitFor} from '@testing-library/react';
import { MemoryRouter, Route, Router } from 'react-router';
import Info, {GET_POKEMON_INFO} from '../pages/Info';
import { MockedProvider } from '@apollo/client/testing';
import renderer from "react-test-renderer";

const mocks = [
    {
      request: {
        query: GET_POKEMON_INFO,
        variables: {
            "where": {
                "id": 50
            }
        },
      },
      result: {
        data: {
          pokemon: { 
              id: 50, 
              pokedexNr: '365', 
              name: 'cola', 
              generation: 1, 
              species: 'brus', 
              type1: 'sukker', 
              type2: 'water', 
              heightMeter: 0.6, 
              weightKg: 0.5, 
              hp: 50, 
              attack: 50, 
              defense: 50, 
              sp_attack: 50, 
              sp_defense: 50, 
              speed: 50, 
              ability1: 'superspeed', 
              ability2: 'high-energy', 
              ability3: 'bad-teeth',
              aggregated_rating: 4
            },
        },
      },
    },
  ];

describe("correct info is displayed", () => {
    beforeEach(async () => {
        /*
        This render function will throw errors about using the function act(). These errors will
        not hinder the tests and can be ignored.
        */
        render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <MemoryRouter initialEntries={["/info/50"]}>
                    <Route path="/info/:id">
                        <Info/>
                    </Route>
                </MemoryRouter>
            </MockedProvider>
        )

        await new Promise(resolve => setTimeout(resolve, 0));
    })

    it("correct pokenum, name and gen", () => {
        expect(screen.getByText(/#365/i)).toHaveClass("pokenum")
        expect(screen.getByText(/cola/i)).toHaveClass("pokename")
        expect(screen.getByText(/Generasjon:/i)).toHaveTextContent("Generasjon: 1")
    })

    it("correct weight and height", () => {
        expect(screen.getByText(/Weight:/i)).toHaveTextContent("0.5")
        expect(screen.getByText(/Height/i)).toHaveTextContent("0.6")
    })

    it("correct total stats", async () => {
        await waitFor(() => expect(screen.getByText(/300/i)).toHaveTextContent("300"))
    })

    it("correct abilities", () => {
        expect(screen.getByText(/1./i)).toHaveTextContent("superspeed")
        expect(screen.getByText(/2./i)).toHaveTextContent("high-energy")
        expect(screen.getByText(/Hidden ability./i)).toHaveTextContent("bad-teeth")
    })
    
})

describe("rendering is correct", () => {
    it("Snapshot-test of page", () => {
        const tree = renderer.create(
        <MockedProvider mocks={mocks} addTypename={false}>
            <MemoryRouter initialEntries={["/info/50"]}>
                <Route path="/info/:id">
                    <Info/>
                </Route>
            </MemoryRouter>
        </MockedProvider>).toJSON
        expect(tree).not.toBeNull()
        expect(tree).toMatchSnapshot()
    })

})


