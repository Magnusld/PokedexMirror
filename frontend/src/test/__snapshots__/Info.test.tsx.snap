// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering is correct Snapshot-test of page 1`] = `[Function]`;
